compiler/rustc_middle/src/ty/layout.rs

```diff
@@ -3359,7 +3370,35 @@ impl<'tcx> LayoutCx<'tcx, TyCtxt<'tcx>> {
                 }
 
                 match arg.layout.abi {
-                    Abi::Aggregate { .. } => {}
+                    Abi::Aggregate { .. } => {
+                        // Pass and return structures up to 2 pointers in size by value,
+                        // matching `ScalarPair`. LLVM will usually pass these in 2 registers
+                        // which is more efficient than by-ref.
+                        let max_by_val_size = Pointer.size(self) * 2;
+                        let size = arg.layout.size;
+
+                        if arg.layout.is_unsized() || size > max_by_val_size {
+                            arg.make_indirect();
+                        } else if unlikely!(self.has_all_float(&arg.layout)) {
+                            // We don't want to aggregate floats as an aggregates of Integer
+                            // because this will hurt the generated assembly (#93490)
+                            //
+                            // As an optimization we want to pass homogeneous aggregate of floats
+                            // greater than pointer size as indirect
+                            if size > Pointer.size(self) {
+                                arg.make_indirect();
+                            }
+                        } else {
```

Comment:
```suggestion
                        } else if size > Pointer.size(self) && unlikely!(self.has_all_float(&arg.layout)) {
                            // We don't want to aggregate floats as an aggregates of Integer
                            // because this will hurt the generated assembly (#93490)
                            //
                            // As an optimization we want to pass homogeneous aggregate of floats
                            // greater than pointer size as indirect
                            arg.make_indirect();
                        } else {
```

Reply:
This is not the same logic as before. Before your suggestion if it was `self.has_all_float(&arg.layout)` but NOT `size > Pointer.size(self)` nothing would happen, with this suggestion it will take the else branch doing exactly the opposite of what we want to make.

Reply:
Also, even if it's desirable that small aggregates with float has a special treatment, it should probably be a separate PR that's backed by separate perf run.

Reply:
Well on `x86_64` this would mean that `f64` -> `i64`, `f32` -> `i32` and `[f32; 2]` -> `i64` which is not ideal and is the reason of this PR, ie let the possibility for the codegen to use floating point register for small aggregate of floats or to put them behind an indirection.

Reply:
`f64` and `f32` are not `Abi::Aggregate` and will not hit this path. For `[f32; 2]`, the current status quo is to convert it to integer.

Essentially this PR is doing two things:
1. Pass not-all-float aggregates smaller than 2 register directly
2. Stop casting all-float aggregates smaller than 1 register to integer

I am suggesting just do 1 in this PR.

Reply:
I tried your suggestion locally and only one test failed. I will push it after the perf run so we can have a baseline to know if it's even necessary or not.

Reply:
Well the perf impact of introducing this "fast path" is negligible: https://github.com/rust-lang/rust/pull/97559#issuecomment-1142418639